internal/v2/s2av2_e2e_test.go

```diff
@@ -256,7 +284,48 @@ func TestEndToEndUsingFakeS2AOnUDSEmptyId(t *testing.T) {
 	// Finally, start up the client.
 	ctx, cancel := context.WithTimeout(context.Background(), defaultE2ETimeout)
 	defer cancel()
-	runClient(ctx, t, clientS2AAddr, serverAddr, nil)
+	runClient(ctx, t, clientS2AAddr, serverAddr, nil, nil)
+}
+
+func TestGRPCFallbackEndToEndUsingFakeS2AOverTCP(t *testing.T) {
+	fallback.FallbackTLSConfigGRPC = tls.Config{
+		MinVersion:         tls.VersionTLS13,
+		ClientSessionCache: nil,
+		NextProtos:         []string{"h2"},
+		// set for testing only
+		InsecureSkipVerify: true,
+	}
+	os.Setenv(accessTokenEnvVariable, "TestE2ETCP_token")
+	// Start the fake S2A for the server.
+	serverS2AAddr := startFakeS2A(t, "TestE2ETCP_token")
+	t.Logf("Fake handshaker for server running at address: %v", serverS2AAddr)
+
+	// Start the server.
+	localIdentities := []*commonpbv1.Identity{
+		{
+			IdentityOneof: &commonpbv1.Identity_Hostname{
+				Hostname: "test_rsa_server_identity",
+			},
+		},
+	}
+	serverAddr := startServer(t, serverS2AAddr, localIdentities)
+	fallbackServerAddr := startFallbackServer(t)
+	t.Logf("server running at address: %v", serverAddr)
+	t.Logf("fallback server running at address: %v", fallbackServerAddr)
+
+	// Finally, start up the client.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultE2ETimeout)
+	defer cancel()
+	fallbackHandshake, err := fallback.DefaultFallbackClientHandshakeFunc(fallbackServerAddr)
+	if err != nil {
+		t.Errorf("error creating fallback handshake function: %v", err)
+	}
+	// Set wrong S2A address for client to trigger S2A failure and fallback.
+	runClient(ctx, t, "not_exist", serverAddr, &commonpbv1.Identity{
```

Comment:
In this test, how do we know the fallback was invoked?

Reply:
I added a wrapper to the fallback function, which flips a bool var to indicate being called.